content/post/postgres/rfc-extension-packaging-lookup.md

```diff
@@ -0,0 +1,496 @@
+│   │   └── pair.html
+│   └── pair.md
+└── share
+    ├── pair--1.0--1.1.sql
+    └── pair--1.1.sql
+```
+
+A binary application like [pg_top] would live in the `pg_top` directory,
+structured something like:
+
+```
+pg_top
+├── HISTORY.rst
+├── INSTALL.rst
+├── LICENSE
+├── README.rst
+├── bin
+│   └── pg_top
+└── doc
+    └── man
+        └── man3
+            └── pg_top.3
+```
+
+And a C extension like [semver] would live in the semver directory and be
+structured something like:
+
+``` tree
+semver
+├── LICENSE
+├── README.md
+├── semver.control
+├── doc
+│   └── semver.md
+├── lib
+│   ├── semver.dylib
+│   └── bitcode
+│       ├── semver
+│       │   └── semver.bc
+│       └── semver.index.bc
+└── share
+    ├── semver--1.0--1.1.sql
+    └── semver--1.1.sql
+```
+
+## Phase Two: Preloading
+
+The above-proposed [solution](#solution) does not allow shared modules
+distributed with extensions to compatibly be loaded via [shared library
+preloading], because extension modules wil no longer live in the
+[`dynamic_library_path`]. Users can specify full paths, however. For example,
+instead of:
+
+``` ini
+shared_preload_libraries = 'pg_partman_bgw'
+```
+
+One could use:
+
+```ini
+shared_preload_libraries = '/opt/postgres/extensions/pg_partman_bgw/lib/pg_partman_bgw'
+```
+
+But users will likely find this pattern cumbersome, especially for extensions
+with multiple shared modules. Perhaps some special syntax could be added, for
+example:
+
+```ini
+shared_preload_libraries = '$extension_path:pg_partman_bgw'
+```
+
+But this overloads the semantics of `shared_preload_libraries` and friends
+rather heavily, not to mention the [`LOAD`] command.
+
+Therefore, as a follow up to the [solution](#solution) proposed above, this
+RFC proposes additional changes to PostgreSQL.
+
+### Extension Preloading
+
+Add new GUCs that complement [shared library preloading], but for *extension*
+module preloading:
+
+*   `shared_preload_extensions`
+*   `session_preload_extensions`
+*   `local_preload_extensions`
+
+Each takes a list of extensions for which to preload shared modules. In
+addition, another new GUC, `local_plugins`, will contain a list of
+administrator-approved extensions users are allowed to include in
+`local_preload_extensions`. This GUC complements [`local_preload_libraries`]'s
+use of a `plugins` directory.
+
+Then modify the preloading code to also preload these files. For each
+extension in a list, it would:
+
+*   Search each `$extension_path` for the extension.
+*   When found, load all the shared libraries in `$extension/lib`.
+
+For example, to load all shared modules in the `pg_partman` extension, set:
+
+```ini
+shared_preload_extensions = 'pg_partman'
+```
+
+To load a single shared module from an extension, give its name after the
+extension name and a slash. This example will load only the `pg_partman_bgw`
+shared module from the `pg_partman` extension:
+
+```ini
+shared_preload_extensions = 'pg_partman/pg_partman_bgw'
```

Comment:
Here the syntax looks like a path, but it isn't: there's a `/lib/` in between, unless I got something wrong. Then I'd either make it a full path (`pg_partman/lib/pg_partman_bgw`) or use some other separator (e.g. `pg_partman::pg_partman_bgw`).

Reply:
Done in 15c4952.